File: MapperCore/Mapper.cs
using System;
using System.Data.Linq;
using System.Linq.Expressions;
using MapperCore.Core;
using MapperCore.Exception;
using System.Linq;
using System.Collections.Generic;
using System.Reflection;

namespace MapperCore
{

    /// <summary>
    /// Class de base pour l'accès au mapping
    /// </summary>
    public static class Mapper
    {
        #region Variables

        private static MapperConfigurationBase current;
        private static Func<Type, object> constructorFunc;

        #endregion

        #region Méthodes publiques

        /// <summary>
        /// Maps the specified source.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <param name="source">The business.</param>
        /// <returns></returns>
        public static TDest Map<TSource, TDest>(TSource source)
            where TSource : class
            where TDest : class
        {
            try
            {
                MapperConfiguration<TSource, TDest> mapper = GetMapper<TSource, TDest>();
                Func<TSource, TDest> query = mapper.GetFuncDelegate();
                if (query != null)
                {
                    TDest result = query(source);
                    //Action à exécutées après le mapping
                    mapper.ExecuteAfterActions(source, result);

                    return result;
                }
            }
            catch (System.Exception ex)
            {

                throw ex;
            }
            return null;
        }

        /// <summary>
        /// Gets the query expression.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <returns></returns>
        public static Expression<Func<TSource, TDest>> GetQueryExpression<TSource, TDest>()
            where TSource : class
            where TDest : class
        {
            return GetMapper<TSource, TDest>().GetLambdaExpression();
        }

        /// <summary>
        /// Creates the map.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <returns></returns>
        public static MapperConfiguration<TSource, TDest> CreateMap<TSource, TDest>()
            where TSource : class
            where TDest : class
        {
            MapperConfigurationBase map = MapperConfigurationContainer.Instance.Find(typeof(TSource), typeof(TDest));
            if (map == null)
            {
                map = new MapperConfiguration<TSource, TDest>();
                MapperConfigurationContainer.Instance.Add(map);
            }
            return map as MapperConfiguration<TSource, TDest>;
        }

        /// <summary>
        /// Indique le service d'injection à utilisée
        /// </summary>
        /// <param name="constructor">The constructor.</param>
        public static void ConstructServicesUsing(Func<Type, object> constructor)
        {
            constructorFunc = constructor;
        }

        /// <summary>
        /// Efface tout les mappeur existants
        /// </summary>
        public static void Reset()
        {
            MapperConfigurationContainer.Instance.Clear();
        }

        /// <summary>
        /// Gets the data load options for linqToSql.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <returns></returns>
        public static DataLoadOptions GetDataLoadOptionsLinq<TSource, TDest>()
            where TSource : class
            where TDest : class
        {
            MapperConfiguration<TSource, TDest> map = GetMapper<TSource, TDest>();
            return map.GetDataLoadOptionsLinq();
        }

        /// <summary>
        /// Initialises les mappeurs.
        /// </summary>
        public static void Initialize()
        {
            MapperConfigurationContainer configRegister = MapperConfigurationContainer.Instance;

            foreach (MapperConfigurationBase mapper in configRegister)
            {
                mapper.CreateMappingExpression(constructorFunc);
            }
        }

        /// <summary>
        /// Gets the query.
        /// </summary>
        /// <typeparam name="TSource">The type of the source.</typeparam>
        /// <typeparam name="TDest">The type of the dest.</typeparam>
        /// <returns></returns>
        public static Func<TSource, TDest> GetQuery<TSource, TDest>()
            where TSource : class
            where TDest : class
        {
            return GetMapper<TSource, TDest>().GetFuncDelegate();
        }
        #endregion

        #region Méthodes privées

        internal static MapperConfiguration<TSource, TDest> GetMapper<TSource, TDest>()
            where TSource : class
            where TDest : class
        {
            if (current == null || (current != null && current.TypeDest.FullName != typeof(TDest).FullName && current.TypeSource.FullName != typeof(TSource).FullName))
                current = GetMapper(typeof(TSource), typeof(TDest));

            return (current as MapperConfiguration<TSource, TDest>);

        }

        internal static MapperConfigurationBase GetMapper(Type tSource, Type tDest)
        {
            MapperConfigurationBase mapConfig = MapperConfigurationContainer.Instance.Find(tSource, tDest);
            if (mapConfig != null)
            {
                return mapConfig;
            }
            else
            {
                throw new NoFoundMapperException(tSource, tDest);
            }
        }
      
       


        #endregion
    }
}

File: VB/Mapper/Mapper.vb
Imports System.Data.Linq
Imports System.Linq.Expressions
Imports MapperExpression.Core
Imports MapperExpression.Exception
Imports System.Reflection


''' <summary>
''' Class de base pour l'accès au mapping
''' </summary>
Public Class Mapper

#Region "Variables"

    Private Shared current As MapperConfigurationBase
    Private Shared constructorFunc As Func(Of Type, Object)

#End Region

#Region "Méthodes publiques"

    ''' <summary>
    ''' Maps the specified source.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <param name="source">The business.</param>
    ''' <returns></returns>
    Public Shared Function Map(Of TSource As Class, TDest As Class)(source As TSource) As TDest
        Try
            Dim mapper = GetMapper(Of TSource, TDest)()

            Dim query = mapper.GetFuncDelegate()

            If query IsNot Nothing Then
                Dim result As TDest = query(source)

                'Action à exécuter après le mapping
                mapper.ExecuteAfterActions(source, result)

                Return result

            End If
        Catch ex As System.Exception

            Throw ex
        End Try
        Return Nothing
    End Function

    ''' <summary>
    ''' Gets the query expression.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function GetQueryExpression(Of TSource As Class, TDest As Class)() As Expression(Of Func(Of TSource, TDest))
        Return GetMapper(Of TSource, TDest)().GetLambdaExpression()
    End Function

    ''' <summary>
    ''' Creates the map.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function CreateMap(Of TSource As Class, TDest As Class)() As MapperConfiguration(Of TSource, TDest)
        Dim map = MapperConfigurationContainer.Instance.Find(GetType(TSource), GetType(TDest))

        If map Is Nothing Then
            map = New MapperConfiguration(Of TSource, TDest)()
            MapperConfigurationContainer.Instance.Add(map)
        End If

        Return TryCast(map, MapperConfiguration(Of TSource, TDest))
    End Function

    ''' <summary>
    ''' Indique le service d'injection à utilisée
    ''' </summary>
    ''' <param name="constructor">The constructor.</param>
    Public Shared Sub ConstructServicesUsing(constructor As Func(Of Type, Object))
        constructorFunc = constructor
    End Sub

    ''' <summary>
    ''' Efface tout les mappeurs existants
    ''' </summary>
    Public Shared Sub Reset()
        MapperConfigurationContainer.Instance.Clear()
    End Sub

    ''' <summary>
    ''' Gets the data load options for linqToSql.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function GetDataLoadOptionsLinq(Of TSource As Class, TDest As Class)() As DataLoadOptions
        Dim map = GetMapper(Of TSource, TDest)()
        Return map.GetDataLoadOptionsLinq()
    End Function

    ''' <summary>
    ''' Initialises les mappeurs.
    ''' </summary>
    Public Shared Sub Initialize()
        Dim configRegister As MapperConfigurationContainer = MapperConfigurationContainer.Instance
        For Each mapper As MapperConfigurationBase In configRegister
            mapper.CreateMappingExpression(constructorFunc)
        Next
    End Sub

    ''' <summary>
    ''' Gets the query.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Public Shared Function GetQuery(Of TSource As Class, TDest As Class)() As Func(Of TSource, TDest)
        Return GetMapper(Of TSource, TDest)().GetFuncDelegate()
    End Function

#End Region

#Region "Méthodes privées"

    ''' <summary>
    ''' Gets the mapper.
    ''' </summary>
    ''' <typeparam name="TSource">The type of the source.</typeparam>
    ''' <typeparam name="TDest">The type of the dest.</typeparam>
    ''' <returns></returns>
    Friend Shared Function GetMapper(Of TSource As Class, TDest As Class)() As MapperConfiguration(Of TSource, TDest)
        If current Is Nothing _
            OrElse (current IsNot Nothing _
                                      AndAlso current.TypeDest.FullName <> GetType(TDest).FullName _
                                      AndAlso current.TypeSource.FullName <> GetType(TSource).FullName) Then
            current = GetMapper(GetType(TSource), GetType(TDest))
        End If

        Return TryCast(current, MapperConfiguration(Of TSource, TDest))

    End Function

    ''' <summary>
    ''' Gets the mapper.
    ''' </summary>
    ''' <param name="tSource">The t source.</param>
    ''' <param name="tDest">The t dest.</param>
    ''' <returns></returns>
    ''' <exception cref="System.Exception">Impossible de trouver la configuration entre le type source + tSource.Name +  et le type destination + tDest.Name</exception>
    Friend Shared Function GetMapper(tSource As Type, tDest As Type) As MapperConfigurationBase
        Dim mapConfig = MapperConfigurationContainer.Instance.Find(tSource, tDest)
        If mapConfig IsNot Nothing Then
            Return mapConfig
        Else
            Throw New NoFoundMapperException(tSource, tDest)
        End If
    End Function

    
#End Region

End Class



